File: src/registry.ts
/**
 * Key registry schema, validation, lookup, and public key decoding.
 *
 * A registry holds an array of key entries that map authorities to their
 * Ed25519 public keys with validity date ranges.
 */

import { base64Decode } from './base64url.js';
import { sanitizeForError } from './credential.js';
import { DATE_RE, isValidDate } from './validation.js';

export type KeyEntry = {
  authority: string;
  from: string;
  to: string | null;
  algorithm: 'Ed25519';
  public_key: string;
  note: string;
};

export type Registry = { keys: KeyEntry[] };

const REGISTRY_FIELDS = new Set<string>(['keys']);

export const MAX_REGISTRY_KEYS = 1000;

const ENTRY_FIELDS = new Set<string>([
  'authority',
  'from',
  'to',
  'algorithm',
  'public_key',
  'note',
]);

/** Ed25519 SPKI DER prefix (12 bytes): OID 1.3.101.112 wrapped in SubjectPublicKeyInfo. */
export const ED25519_SPKI_PREFIX = new Uint8Array([
  0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
]);

function validateEntry(entry: unknown, index: number): KeyEntry {
  if (entry === null || typeof entry !== 'object' || Array.isArray(entry)) {
    throw new Error(`keys[${index}] must be a plain object`);
  }

  const record = entry as Record<string, unknown>;

  // Reject extra fields.
  for (const key of Object.keys(record)) {
    if (!ENTRY_FIELDS.has(key)) {
      throw new Error(`keys[${index}]: unexpected field: ${key}`);
    }
  }

  // authority: non-empty string.
  if (!('authority' in record)) {
    throw new Error(`keys[${index}]: missing required field: authority`);
  }
  if (typeof record.authority !== 'string') {
    throw new Error(`keys[${index}]: authority must be a string`);
  }
  if (record.authority.length === 0) {
    throw new Error(`keys[${index}]: authority must not be empty`);
  }
  // NFC-normalize authority during validation so lookups can use plain comparison.
  record.authority = record.authority.normalize('NFC');

  // from: valid date string.
  if (!('from' in record)) {
    throw new Error(`keys[${index}]: missing required field: from`);
  }
  if (typeof record.from !== 'string') {
    throw new Error(`keys[${index}]: from must be a string`);
  }
  if (!isValidDate(record.from)) {
    throw new Error(`keys[${index}]: invalid date for from: ${sanitizeForError(record.from as string)}`);
  }

  // to: valid date string or null.
  if (!('to' in record)) {
    throw new Error(`keys[${index}]: missing required field: to`);
  }
  if (record.to !== null) {
    if (typeof record.to !== 'string') {
      throw new Error(`keys[${index}]: to must be a string or null`);
    }
    if (!isValidDate(record.to)) {
      throw new Error(`keys[${index}]: invalid date for to: ${sanitizeForError(record.to as string)}`);
    }
  }

  if (typeof record.to === 'string' && record.to < record.from) {
    throw new Error(`keys[${index}]: invalid date range: from (${sanitizeForError(record.from as string)}) is after to (${sanitizeForError(record.to as string)})`);
  }

  // algorithm: exactly 'Ed25519'.
  if (!('algorithm' in record)) {
    throw new Error(`keys[${index}]: missing required field: algorithm`);
  }
  if (record.algorithm !== 'Ed25519') {
    throw new Error(`keys[${index}]: algorithm must be 'Ed25519'`);
  }

  // public_key: non-empty string.
  if (!('public_key' in record)) {
    throw new Error(`keys[${index}]: missing required field: public_key`);
  }
  if (typeof record.public_key !== 'string') {
    throw new Error(`keys[${index}]: public_key must be a string`);
  }
  if (record.public_key.length === 0) {
    throw new Error(`keys[${index}]: public_key must not be empty`);
  }

  // note: string (can be empty).
  if (!('note' in record)) {
    throw new Error(`keys[${index}]: missing required field: note`);
  }
  if (typeof record.note !== 'string') {
    throw new Error(`keys[${index}]: note must be a string`);
  }

  return entry as KeyEntry;
}

/** Validate an unknown value as a Registry, throwing on any violation. */
export function validateRegistry(obj: unknown): Registry {
  if (obj === null || typeof obj !== 'object' || Array.isArray(obj)) {
    throw new Error('registry must be a plain object');
  }

  const record = obj as Record<string, unknown>;

  if (!('keys' in record)) {
    throw new Error('missing required field: keys');
  }
  if (!Array.isArray(record.keys)) {
    throw new Error('keys must be an array');
  }
  if (record.keys.length === 0) {
    throw new Error('keys array must not be empty');
  }
  if (record.keys.length > MAX_REGISTRY_KEYS) {
    throw new Error(`registry exceeds maximum key count (${MAX_REGISTRY_KEYS})`);
  }

  // Reject extra top-level fields.
  // Note: Object.keys does NOT enumerate __proto__ after JSON.parse, so this
  // check cannot catch __proto__ at the registry level. Not a real attack vector
  // since the value is inaccessible via Object.keys iteration.
  for (const key of Object.keys(record)) {
    if (!REGISTRY_FIELDS.has(key)) {
      throw new Error(`unexpected field: ${key}`);
    }
  }

  const entries: KeyEntry[] = [];
  for (let i = 0; i < record.keys.length; i++) {
    entries.push(validateEntry(record.keys[i], i));
  }

  return { keys: entries };
}

/**
 * Find all key entries for a given authority.
 *
 * Both the query and each entry's authority are NFC-normalized before
 * comparison.  Comparison is case-sensitive (intentional — authorities are
 * formal names that should match exactly).
 *
 * Results are returned in registry array order.
 */
export function findKeysByAuthority(
  registry: Registry,
  authority: string,
): KeyEntry[] {
  const normalizedQuery = authority.normalize('NFC');
  // entry.authority is already NFC-normalized during validation.
  return registry.keys.filter(
    (entry) => entry.authority === normalizedQuery,
  );
}

/**
 * Check whether a credential date falls within a key's validity range.
 *
 * Uses lexicographic string comparison on ISO 8601 date strings.
 * `from` is inclusive, `to` is inclusive.  `to: null` means no upper bound.
 */
export function isDateInRange(credentialDate: string, key: KeyEntry): boolean {
  if (!DATE_RE.test(credentialDate)) return false;
  if (credentialDate < key.from) return false;
  if (key.to !== null && credentialDate > key.to) return false;
  return true;
}

/**
 * Decode a key entry's `public_key` field to raw 32-byte Ed25519 key material.
 *
 * Accepts either:
 * - 44-byte DER/SPKI encoding (strips the 12-byte prefix)
 * - 32-byte raw key (used as-is)
 *
 * Throws with a clear message for any other length.
 */
export function decodePublicKey(entry: KeyEntry): Uint8Array {
  const bytes = base64Decode(entry.public_key);

  if (bytes.length === 44) {
    // Verify SPKI prefix.
    for (let i = 0; i < ED25519_SPKI_PREFIX.length; i++) {
      if (bytes[i] !== ED25519_SPKI_PREFIX[i]) {
        throw new Error(
          `decodePublicKey: 44-byte key does not have expected Ed25519 SPKI prefix`,
        );
      }
    }
    return bytes.slice(ED25519_SPKI_PREFIX.length);
  }

  if (bytes.length === 32) {
    return bytes;
  }

  throw new Error(
    `decodePublicKey: unexpected key length ${bytes.length} bytes (expected 32 or 44)`,
  );
}

File: go/core/registry.go
package core

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"

	"golang.org/x/text/unicode/norm"
)

// KeyEntry represents a single key in the registry.
type KeyEntry struct {
	Authority string  `json:"authority"`
	From      string  `json:"from"`
	To        *string `json:"to"`
	Algorithm string  `json:"algorithm"`
	PublicKey string  `json:"public_key"`
	Note      string  `json:"note"`
}

// Registry holds the array of key entries.
type Registry struct {
	Keys []KeyEntry `json:"keys"`
}

// MaxRegistryKeys is the maximum number of key entries allowed in a registry.
// Prevents memory exhaustion from oversized registries (~2 MiB JSON at 1000 entries).
const MaxRegistryKeys = 1000

// SupportedAlgorithm is the only supported key algorithm.
const SupportedAlgorithm = "Ed25519"

// ED25519SpkiPrefix is the 12-byte DER/SPKI header for Ed25519 public keys.
var ED25519SpkiPrefix = []byte{
	0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65,
	0x70, 0x03, 0x21, 0x00,
}

// registryFields are the only allowed top-level fields.
var registryFields = map[string]bool{"keys": true}

// entryFields are the only allowed fields in a key entry.
var entryFields = map[string]bool{
	"authority":  true,
	"from":       true,
	"to":         true,
	"algorithm":  true,
	"public_key": true,
	"note":       true,
}

// ValidateRegistry validates parsed JSON bytes as a key registry.
func ValidateRegistry(data []byte) (Registry, error) {
	// First unmarshal into a raw structure to check fields.
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return Registry{}, fmt.Errorf("registry must be a plain object")
	}

	// Check for required "keys" field.
	keysRaw, ok := raw["keys"]
	if !ok {
		return Registry{}, fmt.Errorf("missing required field: keys")
	}

	// Reject extra top-level fields.
	for key := range raw {
		if !registryFields[key] {
			return Registry{}, fmt.Errorf("unexpected field: %s", key)
		}
	}

	// Unmarshal keys as array of raw messages.
	var rawEntries []json.RawMessage
	if err := json.Unmarshal(keysRaw, &rawEntries); err != nil {
		return Registry{}, fmt.Errorf("keys must be an array")
	}
	if len(rawEntries) == 0 {
		return Registry{}, fmt.Errorf("keys array must not be empty")
	}
	if len(rawEntries) > MaxRegistryKeys {
		return Registry{}, fmt.Errorf("registry has %d keys, exceeding maximum %d", len(rawEntries), MaxRegistryKeys)
	}

	entries := make([]KeyEntry, 0, len(rawEntries))
	for i, rawEntry := range rawEntries {
		entry, err := validateEntry(rawEntry, i)
		if err != nil {
			return Registry{}, err
		}
		entries = append(entries, entry)
	}

	return Registry{Keys: entries}, nil
}

func validateEntry(data json.RawMessage, index int) (KeyEntry, error) {
	// Unmarshal into map to check fields.
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return KeyEntry{}, fmt.Errorf("keys[%d] must be a plain object", index)
	}

	// Reject extra fields.
	for key := range raw {
		if !entryFields[key] {
			return KeyEntry{}, fmt.Errorf("keys[%d]: unexpected field: %s", index, key)
		}
	}

	// authority: non-empty string.
	if _, ok := raw["authority"]; !ok {
		return KeyEntry{}, fmt.Errorf("keys[%d]: missing required field: authority", index)
	}
	var authority string
	if err := json.Unmarshal(raw["authority"], &authority); err != nil {
		return KeyEntry{}, fmt.Errorf("keys[%d]: authority must be a string", index)
	}
	if authority == "" {
		return KeyEntry{}, fmt.Errorf("keys[%d]: authority must not be empty", index)
	}

	// from: valid date string.
	if _, ok := raw["from"]; !ok {
		return KeyEntry{}, fmt.Errorf("keys[%d]: missing required field: from", index)
	}
	var from string
	if err := json.Unmarshal(raw["from"], &from); err != nil {
		return KeyEntry{}, fmt.Errorf("keys[%d]: from must be a string", index)
	}
	if !IsValidDate(from) {
		return KeyEntry{}, fmt.Errorf("keys[%d]: invalid date for from: %s", index, SanitizeForError(from))
	}

	// to: null or valid date string.
	if _, ok := raw["to"]; !ok {
		return KeyEntry{}, fmt.Errorf("keys[%d]: missing required field: to", index)
	}
	var to *string
	// Check if it's null.
	if string(raw["to"]) == "null" {
		to = nil
	} else {
		var toStr string
		if err := json.Unmarshal(raw["to"], &toStr); err != nil {
			return KeyEntry{}, fmt.Errorf("keys[%d]: to must be a string or null", index)
		}
		if !IsValidDate(toStr) {
			return KeyEntry{}, fmt.Errorf("keys[%d]: invalid date for to: %s", index, SanitizeForError(toStr))
		}
		to = &toStr
	}

	// Date range validation.
	if to != nil && *to < from {
		return KeyEntry{}, fmt.Errorf("keys[%d]: invalid date range: from (%s) is after to (%s)", index, SanitizeForError(from), SanitizeForError(*to))
	}

	// algorithm: exactly "Ed25519".
	if _, ok := raw["algorithm"]; !ok {
		return KeyEntry{}, fmt.Errorf("keys[%d]: missing required field: algorithm", index)
	}
	var algorithm string
	if err := json.Unmarshal(raw["algorithm"], &algorithm); err != nil || algorithm != SupportedAlgorithm {
		return KeyEntry{}, fmt.Errorf("keys[%d]: algorithm must be '%s'", index, SupportedAlgorithm)
	}

	// public_key: non-empty string.
	if _, ok := raw["public_key"]; !ok {
		return KeyEntry{}, fmt.Errorf("keys[%d]: missing required field: public_key", index)
	}
	var publicKey string
	if err := json.Unmarshal(raw["public_key"], &publicKey); err != nil {
		return KeyEntry{}, fmt.Errorf("keys[%d]: public_key must be a string", index)
	}
	if publicKey == "" {
		return KeyEntry{}, fmt.Errorf("keys[%d]: public_key must not be empty", index)
	}

	// note: string (can be empty).
	if _, ok := raw["note"]; !ok {
		return KeyEntry{}, fmt.Errorf("keys[%d]: missing required field: note", index)
	}
	var note string
	if err := json.Unmarshal(raw["note"], &note); err != nil {
		return KeyEntry{}, fmt.Errorf("keys[%d]: note must be a string", index)
	}
	if controlCharRE.MatchString(note) {
		return KeyEntry{}, fmt.Errorf("keys[%d]: note contains invalid control characters", index)
	}

	return KeyEntry{
		Authority: norm.NFC.String(authority),
		From:      from,
		To:        to,
		Algorithm: algorithm,
		PublicKey: publicKey,
		Note:      note,
	}, nil
}

// DecodePublicKey decodes a key entry's public_key field to raw 32-byte Ed25519 key material.
// Accepts 44-byte SPKI DER (strips 12-byte prefix) or 32-byte raw.
func DecodePublicKey(entry KeyEntry) ([32]byte, error) {
	var result [32]byte

	if len(entry.PublicKey) > 256 {
		return result, fmt.Errorf("decodePublicKey: public_key exceeds maximum length")
	}

	rawKey, err := base64.StdEncoding.DecodeString(entry.PublicKey)
	if err != nil {
		return result, fmt.Errorf("decodePublicKey: invalid base64: %w", err)
	}

	if len(rawKey) == 44 {
		if !bytes.HasPrefix(rawKey, ED25519SpkiPrefix) {
			return result, fmt.Errorf("decodePublicKey: 44-byte key does not have expected Ed25519 SPKI prefix")
		}
		copy(result[:], rawKey[len(ED25519SpkiPrefix):])
		return result, nil
	}

	if len(rawKey) == 32 {
		copy(result[:], rawKey)
		return result, nil
	}

	return result, fmt.Errorf("decodePublicKey: unexpected key length %d bytes (expected 32 or 44)", len(rawKey))
}

// KeyFingerprint returns the SHA-256 fingerprint of a key entry's public key
// as a hex string. Returns an error if the key cannot be decoded.
func KeyFingerprint(entry KeyEntry) (string, error) {
	raw, err := DecodePublicKey(entry)
	if err != nil {
		return "", err
	}
	hash := sha256.Sum256(raw[:])
	return hex.EncodeToString(hash[:]), nil
}

// FindKeysByAuthority returns all key entries matching the given authority.
// Both query and entry authority are NFC-normalized. Comparison is case-sensitive.
func FindKeysByAuthority(reg Registry, authority string) []KeyEntry {
	normalizedQuery := norm.NFC.String(authority)
	var result []KeyEntry
	for _, entry := range reg.Keys {
		// entry.Authority is already NFC-normalized during validation.
		if entry.Authority == normalizedQuery {
			result = append(result, entry)
		}
	}
	return result
}

// IsDateInRange checks if a credential date falls within a key's validity range.
// Uses lexicographic comparison, inclusive on both ends. to=nil means no upper bound.
// Returns false for malformed dates (callers validate format at input boundaries).
func IsDateInRange(credDate string, key KeyEntry) bool {
	if !dateRE.MatchString(credDate) {
		return false
	}
	if credDate < key.From {
		return false
	}
	if key.To != nil && credDate > *key.To {
		return false
	}
	return true
}
